Ruddarr/Dependencies/API/API.swift

```diff
@@ -0,0 +1,107 @@
+    var lookupMovies: (_ instance: Instance, _ query: String) async throws -> [MovieLookup]
+    var systemStatus: (Instance) async throws -> InstanceStatus
+}
+
+extension API {
+    static var live: Self {
+        .init(fetchMovies: { instance in
+            let url = URL(string: instance.url)!
+                .appending(path: "/api/v3/movie")
+
+            return try await request(url: url, authorization: instance.apiKey)
+        }, lookupMovies: { instance, query in
+            let url = URL(string: instance.url)!
+                .appending(path: "/api/v3/movie/lookup")
+                .appending(queryItems: [.init(name: "term", value: query)])
+
+            return try await request(url: url, authorization: instance.apiKey)
+        }, systemStatus: { instance in
+            let url = URL(string: instance.url)!
+                .appending(path: "/api/v3/system/status")
+
+            return try await request(url: url, authorization: instance.apiKey)
+        })
+    }
+
+    fileprivate static func request<Body: Encodable, Response: Decodable>(
+        method: HTTPMethod = .get,
+        url: URL, authorization: String?,
+        body: Body? = nil,
+        decoder: JSONDecoder = .init(),
+        encoder: JSONEncoder = .init(),
+        session: URLSession = .shared
+    ) async throws -> Response {
+        if !NetworkMonitor.shared.isReachable {
+            throw ApiError.noInternet
+        }
+
+        var request = URLRequest(url: url)
+        request.httpMethod = method.rawValue.uppercased()
+        request.addValue("application/json", forHTTPHeaderField: "Accept")
+        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
+
+        print(url)
+
+        if let body {
+            request.httpBody = try encoder.encode(body)
+        }
+
+        if let authorization {
+            request.addValue("Bearer \(authorization)", forHTTPHeaderField: "Authorization")
+        }
+
+        do {
+            let (json, response) = try await URLSession.shared.data(for: request)
+            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 599
+
+            if statusCode >= 300 {
+                throw ApiError.badStatusCode(statusCode)
+            }
+
+            do {
+                return try decoder.decode(Response.self, from: json)
+            } catch let error {
+                throw ApiError.jsonFailure(error)
+            }
+        } catch let apiError as ApiError {
+            // TODO: personally I'd just stick to idiomatic Swift's untyped errors as they have better ergonomics built in to the language.
+            // But if this is important to you, we can keep using strongly typed errors. In that case, we might consider replacing the
+            // `throws` keyword with Swift's Result type as the return value. I went with idiomatic Swift for my function api until told otherwise.
+            throw apiError // don't rewrap in `.requestFailure`
```

Comment:
@pteasima Let's do whatever you think is best with the errors.

I did the `ValidationError` thing because it was really nice to use with the `.alert()`.

Reply:
Yes I agree that we want strongly typed error enums anytime we're actually using it for app logic. We can still catch a specific error type in those cases. For all other errors that we just display to user, there's no need for a separate type.